sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala

```diff
@@ -1136,4 +1121,27 @@ object SparkSession extends Logging {
       SparkSession.clearDefaultSession()
     }
   }
+
+  /**
+   * Initialize extensions if the user has defined a configurator class in their SparkConf.
+   * This class will be applied to the extensions passed into this function.
+   */
+  private[sql] def applyExtensionsFromConf(conf: SparkConf, extensions: SparkSessionExtensions) {
```

Comment:
How about returning `SparkSessionExtensions` from this method, and modifying the secondary constructor of `SparkSession` as:

```scala
private[sql] def this(sc: SparkContext) {
  this(sc, None, None,
    SparkSession.applyExtensionsFromConf(sc.getConf, new SparkSessionExtensions))
}
```

I'm a little worried whether the order we apply extensions might affect.

Reply:
On second thoughts, we could move the method call to the top of the default constructor?

Reply:
The Default constructor of SparkSession?

Reply:
I thought about this and was worried then about multiple invocations of the extensions
Once every time the SparkSession is cloned

Reply:
Oh, I see, moving to the default constructor was not a good idea.
How about the first suggestion?

Reply:
I see, but in that case, we need to ensure that no injection of extensions is used in the default constructor to avoid initializing without extensions from the conf.

Reply:
Eh .. I think it's okay to have a function and returns that updated extensions.

Reply:
Actually either way looks okay.

Reply:
Updated with replacement then :)